### ISSUE TITLE
Error: PythonPipBuilder:None - Binary validation failed!

### ISSUE THREAD

**Issue by ben-elsen** [awslabs/aws-sam-cli#1237]
### Description

When I use sam build in my Cloud9 environment, everything works but when I do so during the deploy with Amplify Console, I get the error: Error: PythonPipBuilder:None - Binary validation failed!

**Comment by TheSriram** [comment]
This means that either python or pip binary validation has failed. it could be the version of the python picked up does not match what is in the template. But better error messages are in the making here.

https://github.com/awslabs/aws-lambda-builders/pull/118

**Comment by ben-elsen** [comment]
The version of python that was installed didn't match the version in the template. Thanks for pointing that out!

**Comment by Tomaszek** [comment]
The issue started to occur after upgrading SAM CLI from 0.11 to 0.17

SAM CLI 0.11
Python 3.7 installed
In SAM template Runtime: python3.6
Everything worked just fine.

After upgrading to SAM CLI 0.17 I'm getting:
Error: PythonPipBuilder:None - Binary validation failed!

After changing in my template runtime settings to: Runtime: python3.7
Everything worked. 

I have two questions:

1. Why older version of SAM CLI was happily building my application with mismatched runtime setting?
2. Can this be set in SAM CLI 0.17 so it would ignore this mismatch again like the old version?

**Comment by awood45** [comment]
I'm honestly not sure how it was succeeding before - a Python 3.6 function build would require Python 3.6 to be present. If we were accepting the wrong binaries before, you'd probably see issues after deploying to Lambda for a number of Python packages.

So for the questions:
1) I don't know, I would have to try and reproduce this, however...
2) No. Requiring the correct Python version to be present to build is the correct behavior, and behaving otherwise could create hard to debug issues.

Closing this for now, feel free to reopen if there are issues with this error when building with the associated Python version available.

**Comment by rmzelnick** [comment]
I'm having the same issue when using bitbucket pipelines, here's the teardown,

```
+ umask 000
+ GIT_LFS_SKIP_SMUDGE=1 git clone --branch="test" --depth 50 https://x-token-auth:$REPOSITORY_OAUTH_ACCESS_TOKEN@bitbucket.org/$BITBUCKET_REPO_FULL_NAME.git $BUILD_DIR
Cloning into '/opt/atlassian/pipelines/agent/build'...
+ git reset --hard 53f189fd10050d31a3403ee83f4834e3b3626a5d
HEAD is now at 53f189f Merged in feature/env_vars (pull request #20)
+ git config user.name bitbucket-pipelines
+ git config user.email commits-noreply@bitbucket.org
+ git config push.default current
+ git config http.${BITBUCKET_GIT_HTTP_ORIGIN}.proxy http://localhost:29418/
+ git remote set-url origin http://bitbucket.org/$BITBUCKET_REPO_FULL_NAME
+ echo ".bitbucket/pipelines/generated" >> .git/info/exclude
+ chmod 777 $BUILD_DIR
  Downloading https://files.pythonhosted.org/packages/9e/48/2d3cf2acbac9a3761e8427e7cc02ed28640c06cf62d35341f5f98ab5c8ea/aws-sam-translator-1.14.0.tar.gz (105kB)
Collecting chevron~=0.12
  Downloading https://files.pythonhosted.org/packages/a3/1e/6ef58385f54c5bbf70b3b451930015aa2d30b3cd5ae324bba4da77c5b889/chevron-0.13.1-py3-none-any.whl
Collecting click~=7.0
  Downloading https://files.pythonhosted.org/packages/fa/37/45185cb5abbc30d7257104c434fe0b07e5a195a6847506c074527aa599ec/Click-7.0-py2.py3-none-any.whl (81kB)
Collecting serverlessrepo==0.1.9
  Downloading https://files.pythonhosted.org/packages/69/a8/3500d8ecd5e14a94b771e5f78a35313f94232cbde41b37888ce88fd9e07c/serverlessrepo-0.1.9-py2.py3-none-any.whl
Collecting aws-lambda-builders==0.4.0
  Downloading https://files.pythonhosted.org/packages/84/c4/408668c8289ba74f03ed6e8015499248e3990629cb85d97e0f72e4bee421/aws_lambda_builders-0.4.0-py3-none-any.whl (86kB)
Collecting Flask~=1.0.2
  Downloading https://files.pythonhosted.org/packages/d8/94/7350820ae209ccdba073f83220cea1c376f2621254d1e0e82609c9a65e58/Flask-1.0.4-py2.py3-none-any.whl (92kB)
Collecting six~=1.11.0
  Downloading https://files.pythonhosted.org/packages/67/4b/141a581104b1f6397bfa78ac9d43d8ad29a7ca43ea90a2d863fe3056e86a/six-1.11.0-py2.py3-none-any.whl
Collecting boto3>=1.9.56,~=1.9
  Downloading https://files.pythonhosted.org/packages/f6/fa/6397049020b312f71c397fff8d10247c2e49da760e2807af7d21e3c23695/boto3-1.9.253-py2.py3-none-any.whl (128kB)
Collecting docker~=4.0
  Downloading https://files.pythonhosted.org/packages/cc/ca/699d4754a932787ef353a157ada74efd1ceb6d1fc0bfb7989ae1e7b33111/docker-4.1.0-py2.py3-none-any.whl (139kB)
Requirement already satisfied: python-dateutil~=2.6 in /usr/local/lib/python3.8/site-packages (from aws-sam-cli) (2.8.0)
Collecting dateparser~=0.7
  Downloading https://files.pythonhosted.org/packages/82/9d/51126ac615bbc4418478d725a5fa1a0f112059f6f111e4b48cfbe17ef9d0/dateparser-0.7.2-py2.py3-none-any.whl (352kB)
Collecting cookiecutter~=1.6.0
  Downloading https://files.pythonhosted.org/packages/16/99/1ca3a75978270288354f419e9166666801cf7e7d8df984de44a7d5d8b8d0/cookiecutter-1.6.0-py2.py3-none-any.whl (50kB)
Requirement already satisfied: PyYAML~=5.1 in /usr/local/lib/python3.8/site-packages (from aws-sam-cli) (5.1.2)
Collecting chardet<3.1.0,>=3.0.2
  Downloading https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl (133kB)
Collecting certifi>=2017.4.17
  Downloading https://files.pythonhosted.org/packages/18/b0/8146a4f8dd402f60744fa380bc73ca47303cccf8b9190fd16a827281eac2/certifi-2019.9.11-py2.py3-none-any.whl (154kB)
Collecting idna<2.9,>=2.5
  Downloading https://files.pythonhosted.org/packages/14/2c/cd551d81dbe15200be1cf41cd03869a46fe7226e7450af7a6545bfc474c9/idna-2.8-py2.py3-none-any.whl (58kB)
Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.8/site-packages (from requests==2.22.0->aws-sam-cli) (1.25.6)
Collecting jsonschema~=3.0
  Downloading https://files.pythonhosted.org/packages/ce/6c/888d7c3c1fce3974c88a01a6bc553528c99d3586e098eee23e8383dd11c3/jsonschema-3.1.1-py2.py3-none-any.whl (56kB)
Requirement already satisfied: setuptools in /usr/local/lib/python3.8/site-packages (from aws-lambda-builders==0.4.0->aws-sam-cli) (41.4.0)
Requirement already satisfied: wheel in /usr/local/lib/python3.8/site-packages (from aws-lambda-builders==0.4.0->aws-sam-cli) (0.33.6)
Collecting Jinja2>=2.10
  Downloading https://files.pythonhosted.org/packages/65/e0/eb35e762802015cab1ccee04e8a277b03f1d8e53da3ec3106882ec42558b/Jinja2-2.10.3-py2.py3-none-any.whl (125kB)
Collecting Werkzeug>=0.14
  Downloading https://files.pythonhosted.org/packages/ce/42/3aeda98f96e85fd26180534d36570e4d18108d62ae36f87694b476b83d6f/Werkzeug-0.16.0-py2.py3-none-any.whl (327kB)
Collecting itsdangerous>=0.24
  Downloading https://files.pythonhosted.org/packages/76/ae/44b03b253d6fade317f32c24d100b3b35c2239807046a4c953c7b89fa49e/itsdangerous-1.1.0-py2.py3-none-any.whl
Requirement already satisfied: botocore<1.13.0,>=1.12.253 in /usr/local/lib/python3.8/site-packages (from boto3>=1.9.56,~=1.9->aws-sam-cli) (1.12.253)
Requirement already satisfied: s3transfer<0.3.0,>=0.2.0 in /usr/local/lib/python3.8/site-packages (from boto3>=1.9.56,~=1.9->aws-sam-cli) (0.2.1)
Requirement already satisfied: jmespath<1.0.0,>=0.7.1 in /usr/local/lib/python3.8/site-packages (from boto3>=1.9.56,~=1.9->aws-sam-cli) (0.9.4)
Collecting websocket-client>=0.32.0
  Downloading https://files.pythonhosted.org/packages/29/19/44753eab1fdb50770ac69605527e8859468f3c0fd7dc5a76dd9c4dbd7906/websocket_client-0.56.0-py2.py3-none-any.whl (200kB)
Collecting regex
  Downloading https://files.pythonhosted.org/packages/6f/a6/99eeb5904ab763db87af4bd71d9b1dfdd9792681240657a4c0a599c10a81/regex-2019.08.19.tar.gz (654kB)
Collecting tzlocal
  Downloading https://files.pythonhosted.org/packages/ef/99/53bd1ac9349262f59c1c421d8fcc2559ae8a5eeffed9202684756b648d33/tzlocal-2.0.0-py2.py3-none-any.whl
Collecting pytz
  Downloading https://files.pythonhosted.org/packages/e7/f9/f0b53f88060247251bf481fa6ea62cd0d25bf1b11a87888e53ce5b7c8ad2/pytz-2019.3-py2.py3-none-any.whl (509kB)
Collecting future>=0.15.2
  Downloading https://files.pythonhosted.org/packages/3f/bf/57733d44afd0cf67580658507bd11d3ec629612d5e0e432beb4b8f6fbb04/future-0.18.1.tar.gz (828kB)
Collecting poyo>=0.1.0
  Downloading https://files.pythonhosted.org/packages/42/50/0b0820601bde2eda403f47b9a4a1f270098ed0dd4c00c443d883164bdccc/poyo-0.5.0-py2.py3-none-any.whl
Collecting whichcraft>=0.4.0
  Downloading https://files.pythonhosted.org/packages/b5/a2/81887a0dae2e4d2adc70d9a3557fdda969f863ced51cd3c47b587d25bce5/whichcraft-0.6.1-py2.py3-none-any.whl
Collecting jinja2-time>=0.1.0
  Downloading https://files.pythonhosted.org/packages/6a/a1/d44fa38306ffa34a7e1af09632b158e13ec89670ce491f8a15af3ebcb4e4/jinja2_time-0.2.0-py2.py3-none-any.whl
Collecting binaryornot>=0.2.0
  Downloading https://files.pythonhosted.org/packages/24/7e/f7b6f453e6481d1e233540262ccbfcf89adcd43606f44a028d7f5fae5eb2/binaryornot-0.4.4-py2.py3-none-any.whl
Collecting pyrsistent>=0.14.0
  Downloading https://files.pythonhosted.org/packages/b9/66/b2638d96a2d128b168d0dba60fdc77b7800a9b4a5340cefcc5fc4eae6295/pyrsistent-0.15.4.tar.gz (107kB)
Collecting attrs>=17.4.0
  Downloading https://files.pythonhosted.org/packages/a2/db/4313ab3be961f7a763066401fb77f7748373b6094076ae2bda2806988af6/attrs-19.3.0-py2.py3-none-any.whl
Collecting importlib-metadata
  Downloading https://files.pythonhosted.org/packages/f6/d2/40b3fa882147719744e6aa50ac39cf7a22a913cbcba86a0371176c425a3b/importlib_metadata-0.23-py2.py3-none-any.whl
Collecting MarkupSafe>=0.23
  Downloading https://files.pythonhosted.org/packages/b9/2e/64db92e53b86efccfaea71321f597fa2e1b2bd3853d8ce658568f7a13094/MarkupSafe-1.1.1.tar.gz
Requirement already satisfied: docutils<0.16,>=0.10 in /usr/local/lib/python3.8/site-packages (from botocore<1.13.0,>=1.12.253->boto3>=1.9.56,~=1.9->aws-sam-cli) (0.15.2)
Collecting arrow
  Downloading https://files.pythonhosted.org/packages/cc/f5/f618f8423fe674a60461e58683edfeead93f4bbf95b9fca0e9699e765b35/arrow-0.15.2-py2.py3-none-any.whl (40kB)
Collecting zipp>=0.5
  Downloading https://files.pythonhosted.org/packages/74/3d/1ee25a26411ba0401b43c6376d2316a71addcc72ef8690b101b4ea56d76a/zipp-0.6.0-py2.py3-none-any.whl
Collecting more-itertools
  Downloading https://files.pythonhosted.org/packages/45/dc/3241eef99eb45f1def35cf93af35d1cf9ef4c0991792583b8f33ea41b092/more_itertools-7.2.0-py3-none-any.whl (57kB)
Building wheels for collected packages: aws-sam-translator, regex, future, pyrsistent, MarkupSafe
  Building wheel for aws-sam-translator (setup.py): started
  Building wheel for aws-sam-translator (setup.py): finished with status 'done'
  Created wheel for aws-sam-translator: filename=aws_sam_translator-1.14.0-cp38-none-any.whl size=132710 sha256=3cd86b08ed432b44cb9d36ae0dd54d8e37c4e9abf0d74e2cd099916b215d1f7c
  Stored in directory: /root/.cache/pip/wheels/34/a1/31/49ac6bfdd8f2a5b9c1bfb4cca62c8456bf3b5832c116686631
  Building wheel for regex (setup.py): started
  Building wheel for regex (setup.py): finished with status 'done'
  Created wheel for regex: filename=regex-2019.8.19-cp38-cp38-linux_x86_64.whl size=708858 sha256=b81e8db79fee23b04b124d7ead38452eb68f7c07be12ca0e2f2097ab4bd399c3
  Stored in directory: /root/.cache/pip/wheels/90/04/07/b5010fb816721eb3d6dd64ed5cc8111ca23f97fdab8619b5be
  Building wheel for future (setup.py): started
  Building wheel for future (setup.py): finished with status 'done'
  Created wheel for future: filename=future-0.18.1-cp38-none-any.whl size=490417 sha256=0ba3f039f86f553991cbc4a10a02ca8f88972bdcb837855fdc9b3546ac80a30c
  Stored in directory: /root/.cache/pip/wheels/74/bc/50/ae030267a192919b289f84661cfeb5573cf383841c543e8696
  Building wheel for pyrsistent (setup.py): started
  Building wheel for pyrsistent (setup.py): finished with status 'done'
  Created wheel for pyrsistent: filename=pyrsistent-0.15.4-cp38-cp38-linux_x86_64.whl size=133912 sha256=9bf6f5f6e4a4085c82929a05a2cafb87d6adc8b3114eca435c64659dfd394fbb
  Stored in directory: /root/.cache/pip/wheels/bb/46/00/6d471ef0b813e3621f0abe6cb723c20d529d39a061de3f7c51
  Building wheel for MarkupSafe (setup.py): started
  Building wheel for MarkupSafe (setup.py): finished with status 'done'
  Created wheel for MarkupSafe: filename=MarkupSafe-1.1.1-cp38-cp38-linux_x86_64.whl size=33832 sha256=7f10971afa6c4fcfa22032b4066217912734d95316405412d0ea53a5ee7e2006
  Stored in directory: /root/.cache/pip/wheels/f2/aa/04/0edf07a1b8a5f5f1aed7580fffb69ce8972edc16a505916a77
Successfully built aws-sam-translator regex future pyrsistent MarkupSafe
Installing collected packages: chardet, certifi, idna, requests, boto3, six, pyrsistent, attrs, more-itertools, zipp, importlib-metadata, jsonschema, aws-sam-translator, chevron, click, serverlessrepo, aws-lambda-builders, MarkupSafe, Jinja2, Werkzeug, itsdangerous, Flask, websocket-client, docker, regex, pytz, tzlocal, dateparser, future, poyo, whichcraft, arrow, jinja2-time, binaryornot, cookiecutter, aws-sam-cli
  Found existing installation: six 1.12.0
    Uninstalling six-1.12.0:
      Successfully uninstalled six-1.12.0
Successfully installed Flask-1.0.4 Jinja2-2.10.3 MarkupSafe-1.1.1 Werkzeug-0.16.0 arrow-0.15.2 attrs-19.3.0 aws-lambda-builders-0.4.0 aws-sam-cli-0.22.0 aws-sam-translator-1.14.0 binaryornot-0.4.4 boto3-1.9.253 certifi-2019.9.11 chardet-3.0.4 chevron-0.13.1 click-7.0 cookiecutter-1.6.0 dateparser-0.7.2 docker-4.1.0 future-0.18.1 idna-2.8 importlib-metadata-0.23 itsdangerous-1.1.0 jinja2-time-0.2.0 jsonschema-3.1.1 more-itertools-7.2.0 poyo-0.5.0 pyrsistent-0.15.4 pytz-2019.3 regex-2019.8.19 requests-2.22.0 serverlessrepo-0.1.9 six-1.11.0 tzlocal-2.0.0 websocket-client-0.56.0 whichcraft-0.6.1 zipp-0.6.0
(1/19) Purging build-deps (20191022.155559)
(2/19) Purging build-base (0.5-r1)
(3/19) Purging file (5.37-r0)
(4/19) Purging g++ (8.3.0-r0)
(5/19) Purging make (4.2.1-r2)
(6/19) Purging libc-dev (0.7.1-r0)
(7/19) Purging musl-dev (1.1.22-r3)
(8/19) Purging fortify-headers (1.1-r0)
(9/19) Purging gcc (8.3.0-r0)
(10/19) Purging binutils (2.32-r0)
(11/19) Purging libatomic (8.3.0-r0)
(12/19) Purging libgomp (8.3.0-r0)
(13/19) Purging libmagic (5.37-r0)
(14/19) Purging mpc1 (1.1.0-r0)
(15/19) Purging mpfr3 (3.1.5-r1)
(16/19) Purging isl (0.18-r0)
(17/19) Purging gmp (6.1.2-r1)
(18/19) Purging libstdc++ (8.3.0-r0)
(19/19) Purging libgcc (8.3.0-r0)
Executing busybox-1.30.1-r2.trigger
OK: 19 MiB in 36 packages
+ /usr/local/bin/sam build -t templates/dev-template.yaml
	SAM CLI now collects telemetry to better understand customer needs.
	You can OPT OUT and disable telemetry collection by setting the
	environment variable SAM_CLI_TELEMETRY=0 in your shell.
	Thanks for your help!
	Learn More: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-telemetry.html
Building resource 'VendorDisbursementsFunction'
Build Failed
Error: PythonPipBuilder:None - Binary validation failed!
Searching for test report files in directories named [test-results, failsafe-reports, test-reports, surefire-reports] down to a depth of 4
Finished scanning for test reports. Found 0 test report files.
Merged test suites, total number tests is 0, with 0 failures and 0 errors.
```

**Comment by consemk-cg** [comment]
when I changed the Runtime property under "AWS::Serverless::Function" resource to match to the runtime version of my python, then sam build started working. Thanks to @sriram-mv .
